doc/config-generic-update-rollback/SONiC_Generic_Config_Update_and_Rollback_Design.md

```diff
@@ -0,0 +1,867 @@
+
+*ACLs*: Updating ACLs require the following steps:
+- Pushing `acl.json` file to the device that contain the new ACL rules
+- Pushing `minigraph.xml` to the device that contains the new ACL interfaces
+- Execute `sudo acl-loader update full /etc/sonic/acl.json --table_name example_acl`
+
+*DHCP*: Updating DHCP config requires the following steps:
+- Pushing `minigraph.xml` to the device that contains the new ACL interfaces
+- Generating `dhcp_servers` JSON configs from the `minigraph.xml`, and saving it as a temporary file
+- Executing `sudo sonic-cfggen -j /tmp/dhcp.json --write-to-db`
+- Restart `dhcp_relay` service
+
+We have explored [SONiC CLI commands](https://github.com/Azure/sonic-utilities/blob/master/doc/Command-Reference.md) to make configuration changes. These CLI commands result in updates to the ConfigDB which are corresponding to the CLI command executed. For example, the config `vlan add 10` will create a new row in the VLAN table of the ConfigDB. But relying on the CLI commands to do partial update is also not feasible as there is no standard way of showing the config after the update. Setting up a different update mechanism for each part of the config is very time consuming and inefficient.
+
+The other challenge to updating a switch is recoverability via rollback. Rollback needs to be with minimum-disruption e.g. if reverting ACL updates DHCP should not be affected. Currently SONiC has a couple of operations that can be candidates for rollback `config load` and `config reload`.
+
+*config reload <config_db.json>* : This command clears all the contents of the ConfigDB and loads the contents of config_db.json into the ConfigDB. After that all the Docker containers  and Linux services are restarted to establish the user specified configuration state in the config_db.json file.
+
+- Pro's
+  - Assured way of affecting a configuration state change
+- Con's
+  - Brings the links down and resets the forwarding state. This operation is disruptive in nature
+  - Time consuming as it may take 2-3 minutes for all the services to come back online. The time taken may vary based on the switch CPU power.
+- Verdict
+  - Cannot be used as a rollback mechanism
+
+*config load <config_db.json>*: This command loads the contents of config_db.json into the ConfigDB. The updates made to the ConfigDB are additive in nature and thus the new configuration state is a combination of the current running state and the partial configuration state specified by the user in the config_db.json file
+
+- Pro's
+  - Quick way to add new configuration changes
+  - It does not disrupt existing service whose configuration is not being modified. So it is non-disruptive in nature
+- Con's
+  - Can't remove existing configuration and can only be used to add/modify the existing configuration
+- Verdict
+  - Cannot be used as a rollback mechanism
```

Comment:
This was true because we did not have backend code for all networking features, that handles delete OP.  I think, Otherwise, we can use it for rollback for any networking feature related config which supports delete operation at the backend.

Reply:
Yes, we are adding the support to delete OP in this design document.